netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -389,14 +389,15 @@ public void run() {
     super.handlerAdded(ctx);
   }
 
-  private void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers)
+  // @return true if completed without calling respondWithHttpError, false if called that method.
+  private boolean onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers)
       throws Http2Exception {
     try {
       // Connection-specific header fields makes a request malformed. Ideally this would be handled
       // by Netty. RFC 7540 section 8.1.2.2
       if (!DISABLE_CONNECTION_HEADER_CHECK && headers.contains(CONNECTION)) {
         resetStream(ctx, streamId, Http2Error.PROTOCOL_ERROR.code(), ctx.newPromise());
-        return;
+        return false;
```

Comment:
Revert these changes, since they aren't needed with the change to onDataRead()? And then your test will handle both HEADERS(endStream=true) and HEADERS+DATA(endStream=true) combinations.

Reply:
done